source/dem/from_file_insertion.cc

```diff
@@ -0,0 +1,188 @@
+#include <core/utilities.h>
+
+#include <dem/from_file_insertion.h>
+
+
+using namespace DEM;
+
+template <int dim>
+FromFileInsertion<dim>::FromFileInsertion(
+  const DEMSolverParameters<dim> &dem_parameters,
+  const std::vector<std::shared_ptr<Distribution>>
+    &distribution_object_container)
+  : Insertion<dim>(distribution_object_container)
+  , remaining_particles_of_each_type(
+      dem_parameters.lagrangian_physical_properties.number.at(0))
+{
+  // Initializing current inserting particle type
+  this->current_inserting_particle_type = 0;
+
+  file_name = dem_parameters.insertion_info.insertion_particles_file;
+}
+
+// The main insertion function. Insert_global_function is used to insert the
+// particles
+template <int dim>
+void
+FromFileInsertion<dim>::insert(
+  Particles::ParticleHandler<dim>                 &particle_handler,
+  const parallel::distributed::Triangulation<dim> &triangulation,
+  const DEMSolverParameters<dim>                  &dem_parameters)
+{
+  if (remaining_particles_of_each_type == 0 &&
+      this->current_inserting_particle_type !=
+        dem_parameters.lagrangian_physical_properties.particle_type_number - 1)
+    {
+      remaining_particles_of_each_type =
+        dem_parameters.lagrangian_physical_properties.number.at(
+          ++this->current_inserting_particle_type);
+    }
+
+
+  if (remaining_particles_of_each_type > 0)
+    {
+      // Read the input file
+      std::map<std::string, std::vector<double>> particles_data;
+      fill_vectors_from_file(particles_data, this->file_name, ";");
+
+      // Number of particles in the file
+      unsigned int n_total_particles_to_insert = particles_data["p_x"].size();
+
+      // Adjusting the value in case we exceed the maximum number of particle in
+      // the simulation.
+      n_total_particles_to_insert =
+        std::min(remaining_particles_of_each_type, n_total_particles_to_insert);
+
+      // Processors 0 will be the only one inserting particles
+      MPI_Comm communicator = triangulation.get_communicator();
+      auto this_mpi_process = Utilities::MPI::this_mpi_process(communicator);
+      const unsigned int n_particles_to_insert_this_proc =
+        this_mpi_process == 0 ? n_total_particles_to_insert : 0;
```

Comment:
Why not just skip the functions if thread =/= 0?

Reply:
We are doing the same thing for the list insertion...